### PR TITLE
Fix for 409: assert on mask target must have NoData defined

### LINKED ISSUE #409: masking functions should warn or convert if raw cell type
Masking seems "broken" when the data tile that the mask is applied to has no nodata defined (raw celltype)

Queries about nodata cell counts are inconsistent with tile contents of the mask .

The masking functions could:

1. throw assertion error, fail outright if the celltype of the data tile has no nodata;
1. warn the user in logs but continue with the current logic;
1. implicitly convert the raw type to the default nodata value; __OR__
1. rely on a user config to decide among above options

Failing has the downside that it may not be evaluated until other parts of the spark job graph have been worked. REsulting in wasted computation and frustration. 

Warnings may not be seen or noticed by the user ... esp true for users in the python api

Converting the celltype may mark valid data as nodata without the user understanding quite why, impacting the result of analytics.

Even with a user config there is the issue of setting a reasonable default.

### PR DESCRIPTION
**Pull request by vpipkt**
Fixes #409

### CODE REVIEW

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+import org.apache.spark.sql.Encoders
+import org.apache.spark.sql.functions._
+import org.locationtech.rasterframes.expressions.accessors.ExtractTile
+import org.locationtech.rasterframes.stats._
+import org.locationtech.rasterframes.tiles.ProjectedRasterTile
+
+
+import geotrellis.raster.testkit.RasterMatchers
+
+class MaskingFunctionsSpec extends TestEnvironment with RasterMatchers {
```
Comment (vpipkt) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: Note we have refactored the mask related function tests out into this spec. Just because we had RasterFunctionsSpec gettting so big and starting to get disorganized.

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+      val cellTyperesult2 = CellType.noNoDataCellTypes.contains(noNoDataCellType)
+      cellTyperesult2 should be (true)
+
+      val df = Seq(TestData.projectedRasterTile(5, 5, 42, TestData.extent, TestData.crs,
+        noNoDataCellType))
+        .toDF("tile")
+      // if this had NoData defined would be a no-op because the tile is all datacells
+      lazy val result = df.select(rf_mask($"tile", $"tile")).collect()
+      an[AssertionError] should be thrownBy(result)
```
Comment (vpipkt) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: The meat of this PR.

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+    }
+
+    it("should throw if no nodata"){
+      val noNoDataCellType = UByteCellType
+      val cellTypeResult = noNoDataCellType match {
+        case _: NoNoData ⇒ 99
+        case _ ⇒ 0
+      }
+//      cellTypeResult should begreater (0)
+      cellTypeResult should be >(0)
```
Comment (vpipkt) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: these several lines are cruft.

File: core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala [hunk 4]
```diff
@@ -73,6 +73,15 @@ abstract class Mask(val left: Expression, val middle: Expression, val right: Exp
   override protected def nullSafeEval(targetInput: Any, maskInput: Any, maskValueInput: Any): Any = {
     implicit val tileSer = TileUDT.tileSerializer
     val (targetTile, targetCtx) = tileExtractor(targetExp.dataType)(row(targetInput))
+
+    // Which of these is preferable? companion object Seq contains or pattern match on trait?
```
Comment (vpipkt) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala]: @metasim what do you prefer here?

Comment (metasim) [reply core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala]: Definitely not `contains`... too wordy. You could also do `targetTile.cellType.isInstanceOf[NoNoData]`.

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala (hunk 1 above)
Comment (metasim) [reply core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: Put this in the `functions` package. I've continued down this vein in https://github.com/locationtech/rasterframes/pull/430

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala (hunk 2 above)
Comment (metasim) [reply core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: I prefer:

```scala
an[AssertionError] should be thrownBy {
  df.select(rf_mask($"tile", $"tile")).collect()
}
```

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+        .withColumn("qa_fill", rf_local_extract_bits($"mask", lit(0)))
+        .withColumn("qa_sat", rf_local_extract_bits($"mask", lit(2), lit(2)))
+        .withColumn("qa_cloud", rf_local_extract_bits($"mask", lit(4)))
+        .withColumn("qa_cconf", rf_local_extract_bits($"mask", 5, 2))
+        .withColumn("qa_snow", rf_local_extract_bits($"mask", lit(9), lit(2)))
+        .withColumn("qa_circonf", rf_local_extract_bits($"mask", 11, 2))
+
+      def checker(colName: String, valFilter: Int, assertValue: Int): Unit = {
+        // print this so we can see what's happening if something  wrong
+        println(s"${colName} should be ${assertValue} for qa val ${valFilter}")
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: Send to `logger.debug` instead.

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        /**  in this unit test setup, the `val` column is an integer that the entire row's mask is full of
+          * filter for the maskValueFilter
+          * then check the columnName and look at the masked data tile given by `columnName`
+          * assert that the `columnName` tile is / is not all nodata based on `resultIsNoData`
+          * */
+
+        val printOutcome = if (resultIsNoData) "all NoData cells"
+        else "all data cells"
+
+        println(s"${columnName} should contain ${printOutcome} for qa val ${maskValueFilter}")
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: `logger.debug`

File: core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        val resultDf = result
+          .filter($"val" === lit(maskValueFilter))
+
+        val resultToCheck: Boolean = resultDf
+          .select(rf_is_no_data_tile(col(columnName)))
+          .first()
+
+        val dataTile = resultDf.select(col(columnName)).as[ProjectedRasterTile].first()
+        println(s"\tData tile values for col ${columnName}: ${dataTile.toArray().mkString(",")}")
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/MaskingFunctionsSpec.scala]: `logger.debug`

File: docs/src/main/paradox/release-notes.md [hunk 8]
```diff
@@ -8,6 +8,7 @@
 * _Breaking_: `rf_spatial_index` renamed `rf_xz2_index` to differentiate between XZ2 and Z2 variants.
 * Added `withSpatialIndex` to RasterSourceDataSource to pre-partition tiles based on tile extents mapped to a Z2 space-filling curve 
 * Add `rf_mask_by_bit`, `rf_mask_by_bits` and `rf_local_extract_bits` to deal with bit packed quality masks. Updated the masking documentation to demonstrate the use of these functions.
+* Throw an `AssertionError` when attempting to apply a mask to a `Tile` whose `CellType` has no NoData defined. ([#409](https://github.com/locationtech/rasterframes/issues/384))
```
Comment (metasim) [inline docs/src/main/paradox/release-notes.md]: I'm questioning whether `AssertionError` is the right one... `require` uses `IllegalArgumentException`, so I think this is more in line with conventional semantics.

File: core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala [hunk 9]
```diff
@@ -73,6 +73,15 @@ abstract class Mask(val left: Expression, val middle: Expression, val right: Exp
...
+//    assert(!CellType.noNoDataCellTypes.contains(targetTile.cellType), maskErrorStr)
+    targetTile.cellType match {
+      case _: NoNoData ⇒ assert(false,
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala]: See comment below about `AssertionError`. The convention I see in Scala is that `require` is more common for user-sourced errors, and `assert` is for trapping coder-sourced errors. So I think `assert` should be changed to `require`.

File: docs/src/main/paradox/release-notes.md (hunk 8 above)
Comment (vpipkt) [reply docs/src/main/paradox/release-notes.md]: I think you are right.
